Shifting to JAX (WIP)

b-biswas opened the pull request:
- [x] Use a name tuple for observation instead of ngmix
- [x] jax.jit metacal.py functions 
- [ ] jax.jit metadetect.py functions (maybe in a different PR?)
- [x] Update test suite according to the new API 
- [ ] Validate results (ongoing)

beckermr committed:
Initial commit

beckermr committed:
ENH set things up (#1)

* ENH set things up

* BUG fix pre-commit

* BUG use correct name for license

* BUG import version

* BUG wrong test path

* DOC add badges

* REF try code cov gha

* DOC add final badge

* BUG produce xml report too

* BUG produce xml report too

beckermr committed:
ENH move over code from deep-metacal (#2)

* ENH move over pixel covariance utils

* ENH use seeded RNGs

* DEP do not need this test anymore

* ENH add deep-metacal stuff and tests for noise handling

* BUG make sure to ignore generated version file

* BUG wrong keyword for width

* PROD ignore numba code

* ENH redo shear measurement to capture more things

* PROD try to adjust coverage

* PROD try to adjust coverage again

* BUG make sure no cover works properly for coverage

* REF use pytest-skip-slow

* REF use pytest-skip-slow

* ENH add tests for deep metacal

* TST add simple smoke test

* TST cancel multiple test jobs

* BUG fix test for deep metacal

* PROD adjust coverage for slow tests

* TST added test of terms

* TST added test of terms

* TST put back low noise for test

* TST put back low noise for test

* STY lint for more things

* TST added test of object noise values

* TST add tests for utils

* STY blacken

* ENH swap12 for estimating m/c

* TST fix test coverage

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#3)

beckermr committed:
ENH add detection functions (#4)

* ENH add detection functions

* STY blacken

* TST try this to get tests to pass

* TST added tests for detection

* BUG use token for codecov

* TST add smoke test for detections

* TST slowly improve tests for obs

* BUG fix bug in subobs

* TST test psf and jacobian

* TST finish tests

* TST make sure images of detections work ok

* TST make sure coverage is ok

* TST adjust test coverage

beckermr committed:
DOC added some docs (#5)

beckermr committed:
ENH rename (#6)

beckermr committed:
ENH add deep-field mdet (#7)

* WIP add mdet

* Apply suggestions from code review

* BUG make sure to fill T

* TST add test of full thing

beckermr committed:
TST enable slow tests (#8)

beckermr committed:
ENH keep bmask flags (#9)

* ENH keep bmask flags

* REF remove white-space changes

* REF put flags in mdet and add test

beckermr committed:
TST make sure slow tests can run

beckermr committed:
ENH add mfrac (#10)

* ENH add mfrac

* BUG make sure to only use mfrac from the image, not the noise

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#11)

beckermr committed:
TST print info from tests

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#12)

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.4.3 → v0.4.7](https://github.com/astral-sh/ruff-pre-commit/compare/v0.4.3...v0.4.7)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

beckermr committed:
test: break up slow tests (#13)

beckermr committed:
fix: make sure slow tests have unique names

beckermr committed:
fix: make sure to cancel old tests

beckermr committed:
fix: already had that

beckermr committed:
test: faster slow tests for cases that should fail (#14)

* test: faster slow tests for cases that should fail

* Update test_metadetect.py

beckermr committed:
test: fix test exiting early

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#16)

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#17)

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#18)

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#19)

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#20)

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#21)

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#22)

b-biswas committed:
Add NT Observation

b-biswas committed:
jaxify metacal func

b-biswas committed:
ngmix obs to NT

b-biswas committed:
to jax

b-biswas committed:
added old version

b-biswas committed:
update tests

b-biswas committed:
updated environment.yml file

b-biswas committed:
pre-commit changes

b-biswas committed:
set jax x64

b-biswas committed:
jax x64

b-biswas committed:
update tests

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#24)

b-biswas committed:
undo changes in utils

b-biswas committed:
revert old code

b-biswas committed:
add jax code

b-biswas committed:
minor [skip ci]

beckermr commented on .github/workflows/tests-slow.yml:
```diff
@@ -91,6 +91,11 @@ jobs:
+        run: |
+          export JAX_ENABLE_X64=True
+          pytest
+
```
Why do we need this extra test block?

beckermr commented on deep_field_metadetect/jaxify/observation.py:
```diff
@@ -0,0 +1,131 @@
+
+
+@jax.tree_util.register_pytree_node_class
+class NTObservation(NamedTuple):
```
Is the goal here to make a JAX-compatible copy of the ngmix observation or maybe should we restrict this class to only the items we need for running deep-field metadetect?

b-biswas replied:
Yes, I agree on that we should remove the unnecessary items. But I was trying ot keep it as close to ngmix as possible until we remove all dependency from ngmix (Gaussian moments, etc...). So maybe for now we can continue with the current class?

beckermr replied:
Yep that's fine.

beckermr commented on deep_field_metadetect/jaxify/observation.py:
```diff
@@ -0,0 +1,131 @@
+        return True
+
+
+def ngmix_Obs_to_NT(obs: ngmix.observation.Observation) -> NTObservation:
```
```suggestion
def ngmix_obs_to_dfmd_obs(obs: ngmix.observation.Observation) -> NTObservation:
```

beckermr commented on deep_field_metadetect/jaxify/observation.py:
```diff
@@ -0,0 +1,131 @@
+
+
+@jax.tree_util.register_pytree_node_class
+class NTObservation(NamedTuple):
```
```suggestion
class DFMdetObservation(NamedTuple):
```

beckermr commented on deep_field_metadetect/jaxify/observation.py:
```diff
@@ -0,0 +1,131 @@
+    )
+
+
+def NT_to_ngmix_obs(nt_obs) -> Observation:
```
```suggestion
def dfmd_obs_to_ngmix_obs(nt_obs) -> Observation:
```

beckermr commented on deep_field_metadetect/utils.py:
```diff
@@ -584,8 +599,10 @@ def make_simple_sim(
+    dim_psf=53,
     buff=26,
     obj_flux_factor=1,
+    return_NT=False,
```
```suggestion
    return_dfmd_obs=False,
```

beckermr commented on environment.yml:
```diff
@@ -16,6 +16,9 @@ dependencies:
   - ngmix
   - numba
   - numpy
+  - dm-tree
```
I cannot find this dep used anywhere. Did I miss it?

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+        raise RuntimeError("Shear value '%s' not regonized!" % shear)
+
+
+# TODO: what should be the value to nxy?
```
Is this still a todo item, or are we OK with a default of 53?

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+from deep_field_metadetect.jaxify.observation import NT_to_ngmix_obs, NTObservation
+
+DEFAULT_SHEARS = ("noshear", "1p", "1m", "2p", "2m")
+DEFAULT_STEP = 0.01
```
Maybe we should move these to a global defaults module shared by both the jax and non-jax versions?

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+    psf : galsim object
+        The PSF.
+    flux : float
+        The output flux of the PSF. Defaults to 1.
```
The doc string needs updating.

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+    reconv_psf : galsim object
+        The reconvolution PSF.
+    sigma : float
+        The width of the reconv PSF befor dilation.
```
The return docstring doesn't match the function. This might be an issue in the non-jax version too. I will fix that if needed. We should fix it here too.

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+    mc_psf_d = jax_get_gauss_reconv_psf_galsim(psf_d, dk_d, nxy_psf, step=step)
+
+    # fwhm_w = jnp.asarray(mc_psf_w.fwhm)
+    # fwhm_d = jnp.asarray(mc_psf_d.fwhm)
```
Do we need this code or can we remove this comment?

Also, do we need the jnp.asarray calls in the `conda` call below?

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+
+    pim = reconv_psf.drawImage(
+        nx=53,
+        ny=53,
```
These need to be set to nxy_psf, right?

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+
+    ns = ns.withGSParams(
+        minimum_fft_size=dims[0] * 4,
+        maximum_fft_size=dims[0] * 4,
```
We'd formally want something like the maximum of the dims here.

Also, do we want to do something like this

```python
with jax.ensure_compile_time_eval():
    minmax_fftsize = max(*dims) * 4
```

and then use `minmax_fftsize` in the calls below?

My concern is that we may need to provide a hint to jax that it should evaluate the array sizes ahead of time. On the other hand, JAX might figure this out because the `dims` are marked as static? I am not sure.

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+    new_mfrac = None
+    new_meta_data = {}
+
+    if obs1.has_bmask() or obs2.has_bmask():
+        new_bmask = _extract_attr(obs1, "bmask", np.int32) | _extract_attr(
+            obs2, "bmask", jnp.int32
+        )
+
+    if obs1.has_ormask() or obs2.has_ormask():
+        new_ormask = _extract_attr(obs1, "ormask", np.int32) | _extract_attr(
+            obs2, "ormask", jnp.int32
+        )
+
+    if obs1.has_noise() or obs2.has_noise():
+        new_noise = _extract_attr(obs1, "noise", np.float32) + _extract_attr(
+            obs2, "noise", jnp.float32
+        )
```
Do we need to use jax.numpy dtypes (i.e. `jnp.float32` or w/e)?

b-biswas replied:
The deault in _extract_attr is now updated

beckermr commented on deep_field_metadetect/jaxify/jax_metacal.py:
```diff
@@ -0,0 +1,492 @@
+        "return_noshear_deep",
+    ],
+)
+def jax_helper_metacal_wide_and_deep_psf_matched(
```
This function looks like it should be private (i.e., have the name `_jax_helper_metacal_wide_and_deep_psf_matched`) since we do not expect people to use it and instead we expect them to call it via the function below.

beckermr commented on deep_field_metadetect/jaxify/tests/test_jax_deep_metacal.py:
```diff
@@ -0,0 +1,339 @@
+        obs_w,
+        obs_d,
+        obs_dn,
+        dk_w=2 * jnp.pi / (53 * 0.2) / 4,
+        dk_d=2 * jnp.pi / (53 * 0.2) / 4,
```
Can you comment on what the various factors here are? Is the factor of 53 related to the PSF size or the image size or do they happen to be the same?

beckermr commented on deep_field_metadetect/jaxify/tests/test_jax_deep_metacal.py:
```diff
@@ -0,0 +1,339 @@
+    if use_mcal:
+        mcal_res = jax_metacal_op_shears(
+            obs_w,
+            dk=jnp.pi / (53 * 0.2) / 4,
+        )
+        for key, value in mcal_res.items():
+            mcal_res[key] = NT_to_ngmix_obs(value)
+    else:
+        mcal_res = jax_metacal_wide_and_deep_psf_matched(
+            obs_w,
+            obs_d,
+            obs_dn,
+            dk_w=jnp.pi / (53 * 0.2) / 4,
+            dk_d=jnp.pi / (53 * 0.2) / 4,
```
These factors for dk don't have the factor of `2` included above. Why do they differ?

beckermr commented on deep_field_metadetect/jaxify/tests/test_jax_metacal.py:
```diff
@@ -0,0 +1,125 @@
+        deep_psf_fac=1.0,
+        return_NT=True,
+    )
+    # jax_gal_psf = get_jax_galsim_object_from_NT_obs_nopix(obs_plus.psf)
```
We should remove comment code if we do not need it.

beckermr commented on deep_field_metadetect/jaxify/tests/test_jax_metadetect.py:
```diff
@@ -0,0 +1,320 @@
+        obs_w,
+        obs_d,
+        obs_dn,
+        dk_w=2 * jnp.pi / (53 * 0.2) / 4,
+        dk_d=2 * jnp.pi / (53 * 0.2) / 4,
+        nxy=201,
+        nxy_psf=53,
```
Again I don't follow these computations.

beckermr reviewed (commented):
I left some comments and questions. A few other items.

1. It might be good to add a series of unit tests that ensure that each function in the non-jax code that has a counterpart in the jax code gives the same answer.
2. It appears that `dk` is compute from image sizes in various places? If yes, then we should be able to compute that as a compile-time constant and remove the need for us to specify it by hand.

b-biswas committed:
initial pr-riv

b-biswas committed:
[skip ci] add stepk computation

b-biswas committed:
[skip ci] remove jnp.array calls

b-biswas committed:
[skip ci] update documentation

beckermr committed:
Merge branch 'main' into to-jax

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#26)

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.9.9 → v0.11.4](https://github.com/astral-sh/ruff-pre-commit/compare/v0.9.9...v0.11.4)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

b-biswas committed:
Bug fix

b-biswas committed:
compare jax and ngmix

b-biswas committed:
fix tests

b-biswas committed:
minor

b-biswas committed:
update tests

b-biswas committed:
use AffineTransform

b-biswas committed:
minor

beckermr commented on deep_field_metadetect/jaxify/observation.py:
```diff
@@ -15,13 +14,9 @@ class DFMdetObservation(NamedTuple):
     ormask: Optional[jax.Array]
     noise: Optional[jax.Array]
-    jacobian: Optional[jax.Array]
+    aft: Optional[jax_galsim.wcs.AffineTransform]
```
Looking ahead, I think we should call this `wcs`.

b-biswas replied:
done!

b-biswas committed:
bug fix

b-biswas committed:
fix bug in testing

b-biswas committed:
minor changes

b-biswas committed:
minor

b-biswas committed:
minor

b-biswas committed:
minor

b-biswas committed:
[skip ci] deconvolution

b-biswas committed:
Gaussmom (#25)

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#27)

beckermr committed:
Merge branch 'main' into to-jax

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#28)

b-biswas committed:
minimal example - deconv bug

b-biswas committed:
for dk to be same as jax for now

b-biswas committed:
fix now for consistency

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#29)

b-biswas committed:
bug fix

b-biswas committed:
updated tests

b-biswas committed:
[skip ci] pre-commit updated

beckermr reviewed (changes requested):
I think defaults of None will be better. Is there a reason you used 0?

b-biswas committed:
[skip ci] change default of min_max_fft to None

b-biswas commented:
> I think defaults of None will be better. Is there a reason you used 0?

For the force_stepk and force_maxk, I was sticking to the Galsim defaults of 0.
But I just switched the max_min_fft_size to default None

b-biswas committed:
minor

b-biswas committed:
fix k values in jax

b-biswas committed:
bug

b-biswas committed:
minor

b-biswas committed:
Merge branch 'main' into to-jax

beckermr commented:
Ohhh interesting. I guess galsim takes a max with the default?

beckermr commented:
I think we want to use the min max fft size everywhere, even if the psf reconv routines?

b-biswas committed:
bug fix

b-biswas committed:
adding max min fft size for reconv

b-biswas committed:
minor

b-biswas committed:
bug fix

b-biswas committed:
fix dk and kim_size in reconv psf

b-biswas committed:
vectorize

b-biswas committed:
minor

b-biswas committed:
text jax and non-jax intermediates

b-biswas committed:
minor

b-biswas committed:
remove black (conflict with ruff)

b-biswas committed:
minor

b-biswas committed:
mark slow tests

b-biswas committed:
Removed notebooks
